File: Julia/MergeSort.jl
function merge(left, right)
    result = []
    i = 1
    j = 1

    while i <= length(left) && j <= length(right)
        if left[i] < right[j]
            push!(result, left[i])
            i += 1
        else
            push!(result, right[j])
            j += 1
        end
    end

    while i <= length(left)
        push!(result, left[i])
        i += 1
    end

    while j <= length(right)
        push!(result, right[j])
        j += 1
    end

    return result
end

function merge_sort(arr)
    if length(arr) <= 1
        return arr
    end

    mid = div(length(arr), 2)
    left = merge_sort(arr[1:mid])
    right = merge_sort(arr[mid+1:end])

    return merge(left, right)
end

File: Go/merge_sort.go
package main

func MergeSort(a []int) []int {
	if len(a) <= 1 {
		return cloneLocal(a)
	}

	mid := len(a) / 2
	left := MergeSort(a[:mid])
	right := MergeSort(a[mid:])

	return merge(left, right)
}

func merge(left, right []int) []int {
	out := make([]int, 0, len(left)+len(right))
	i, j := 0, 0

	for i < len(left) && j < len(right) {
		if left[i] <= right[j] {
			out = append(out, left[i])
			i++
		} else {
			out = append(out, right[j])
			j++
		}
	}
	out = append(out, left[i:]...)
	out = append(out, right[j:]...)
	return out
}

func cloneLocal(a []int) []int {
	out := make([]int, len(a))
	copy(out, a)
	return out
}